### ISSUE TITLE
Dimmer

### ISSUE THREAD

**Issue by ahutchings** [MisterWil/abodepy#22]
This is the JSON for a Z-Wave dimmer ([GE Z-Wave Plus In-Wall Smart Dimmer](https://byjasco.com/products/ge-z-wave-plus-wall-smart-dimmer)) connected through Abode.

I am able to control the on/off state through HASS currently (the dimmer shows up as a light), but not the dimmer percentage.

```json
{
  "id": "ZW:0000000b",
  "type_tag": "device_type.dimmer",
  "type": "Dimmer",
  "name": "West Deck Light",
  "area": "1",
  "zone": "10",
  "sort_order": "",
  "is_window": "",
  "bypass": "0",
  "schar_24hr": "0",
  "sresp_24hr": "0",
  "sresp_mode_0": "0",
  "sresp_entry_0": "0",
  "sresp_exit_0": "0",
  "group_name": "Lights",
  "group_id": "<snip>",
  "default_group_id": "1",
  "sort_id": "1",
  "sresp_mode_1": "0",
  "sresp_entry_1": "0",
  "sresp_exit_1": "0",
  "sresp_mode_2": "0",
  "sresp_entry_2": "0",
  "sresp_exit_2": "0",
  "sresp_mode_3": "0",
  "uuid": "<snip>",
  "sresp_entry_3": "0",
  "sresp_exit_3": "0",
  "sresp_mode_4": "0",
  "sresp_entry_4": "0",
  "sresp_exit_4": "0",
  "version": "",
  "origin": "abode",
  "has_subscription": null,
  "control_url": "api/v1/control/light/ZW:0000000b",
  "deep_link": null,
  "status_color": "#5cb85c",
  "faults": {
    "low_battery": 0,
    "tempered": 0,
    "supervision": 0,
    "out_of_order": 0,
    "no_response": 0
  },
  "status": "Off",
  "statuses": {
    "saturation": "N/A",
    "hue": "N/A",
    "level": "0",
    "switch": "0",
    "color_temp": "N/A",
    "color_mode": "N/A"
  },
  "status_ex": "",
  "actions": [
    { "label": "Switch off", "value": "a=1&z=10&sw=off;" },
    { "label": "Switch on", "value": "a=1&z=10&sw=on;" },
    { "label": "Toggle", "value": "a=1&z=10&sw=toggle;" },
    { "label": "0%", "value": "a=1&z=10&sw=0;" },
    { "label": "10%", "value": "a=1&z=10&sw=10;" },
    { "label": "20%", "value": "a=1&z=10&sw=20;" },
    { "label": "30%", "value": "a=1&z=10&sw=30;" },
    { "label": "40%", "value": "a=1&z=10&sw=40;" },
    { "label": "50%", "value": "a=1&z=10&sw=50;" },
    { "label": "60%", "value": "a=1&z=10&sw=60;" },
    { "label": "70%", "value": "a=1&z=10&sw=70;" },
    { "label": "80%", "value": "a=1&z=10&sw=80;" },
    { "label": "90%", "value": "a=1&z=10&sw=90;" },
    { "label": "100%", "value": "a=1&z=10&sw=99;" }
  ],
  "status_icons": [],
  "statusEx": "0",
  "icon": "assets/icons/bulb-1.svg"
}
```

**Comment by shred86** [comment]
I'm new to Python but I've started digging around to try to get dimmers working. I have it working on my Home Assistant setup with some changes to this abodepy python library. The changes I made are first to the light.py file. When has_brightness is called, it's trying to read 'statusEx' from Abode's device JSON file which appears to be empty unless the light is actually on. Even if it's on, 'has_brightness' doesn't appear to be working properly. I basically changed this:

`    @property
    def is_dimmable(self):
        """Device is dimmable."""
        return self.has_brightness`

To this:

`    @property
    def is_dimmable(self):
        """Device is dimmable."""
        if self._json_state.get('type') == 'Dimmer':
            print('Device is a dimmer')
            return True`

This way if the device is a dimmer according to the Abode device JSON, it will return True. All of my dimmers are now showing up in Home Assistant with a dimmer control and supported_features set to 1 (before everything was set to 0).

This change alone allows me to set the brightness from HA using the 'brightness' key, but the issue now is when using the HA brightness slider, it's setting a value from 0 to 255 where Abode actually uses 0-100. To fix this, I simply changed this:

`            level_data = {
                'level': str(level)
            }`

To this:

`            level_data = {
                'level': int(level / 2.55)
            }`

Which simply converts the 0-255 value into a 0-100 integer. However, looking at the HA logs, when I set the brightness I get an error message which states:

`2018-03-27 22:29:58 ERROR (MainThread) [homeassistant.core] Error executing service <ServiceCall light.turn_on: entity_id=['light.island_light'], brightness=81>
Traceback (most recent call last):
  File "/Users/myuser/virtualenvs/homeassistant/lib/python3.6/site-packages/homeassistant/core.py", line 1006, in _event_to_service_call
    await service_handler.func(service_call)
  File "/Users/myuser/virtualenvs/homeassistant/lib/python3.6/site-packages/homeassistant/components/light/__init__.py", line 344, in async_handle_light_service
    await light.async_turn_on(**params)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/futures.py", line 327, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/tasks.py", line 250, in _wakeup
    future.result()
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/asyncio/futures.py", line 243, in result
    raise self._exception
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/myuser/virtualenvs/homeassistant/lib/python3.6/site-packages/homeassistant/components/light/abode.py", line 51, in turn_on
    self._device.set_level(kwargs[ATTR_BRIGHTNESS])
  File "/Users/myuser/virtualenvs/homeassistant/lib/python3.6/site-packages/abodepy/devices/__init__.py", line 81, in set_level
    raise AbodeException((ERROR.SET_STATUS_STATE))
abodepy.exceptions.AbodeException: Device status/level value does not match request value
`

Still trying to figure what's causing this. The brightness slider in HA is also not being updated with the correct brightness setting when the light is initially turned on.

Anyways, I'm brand new to Python and programming so there's likely a better way to do all of this. I'll continue to work on this and if I get everything working, I'll see if I can submit these changes to be added to the abodepy.

**Comment by ahutchings** [comment]
Thanks, I see the dimmer control in HA now! Closing this issue.

**Comment by shred86** [comment]
The updated Home Assistant Abode light component isn't included in HA 0.67 but should be in the next release (it's in the dev build right now). While dimmers will show up, just realize the code to convert the HA brightness value to what Abode uses isn't there yet so the brightness sliders might be a little off in HA right now.
